### PR TITLE
feat: skip past task occurrences when enabling a disabled task

### PR DESCRIPTION
**Pull request by shubhamms**
Add TaskEnableOptions as an optional parameter to be used to modify task when enabling a task. This PR only gives recomputeNextRunTime as an option to recompute the next run time of a disabled scheduled task while enabling it

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -1,3 +1,8 @@
+### Features
+
+# [1.3.0](https://github.com/princjef/irontask/compare/v1.2.0...v1.3.0) (2020-09-10)
```
Comment (princjef) [inline CHANGELOG.md:3]: please revert the changes in this file. the build system will automatically populate this from your commit message

File: package.json [hunk 2]
```diff
@@ -1,6 +1,6 @@
 {
   "name": "irontask",
-  "version": "1.2.0",
+  "version": "1.3.0",
```
Comment (princjef) [inline package.json:3]: same in this and the package-lock. the version bump is done automatically by the build system

File: src/types/public.ts [hunk 3]
```diff
@@ -55,3 +55,14 @@ export interface TimeoutsOptions {
...
+
+export interface TaskEnablingOptions {
+  /**
+   * Enable a recurring task while skipping all the past occurrences
+   *
+   * @defaultValue false
+   *
+   * @public
+   */
+  skipPastOccurrences?: boolean;
```
Comment (princjef) [inline src/types/public.ts:67]: what would you think about naming this `recomputeSchedule` or something like it instead? if memory serves, the default behavior isn't to run every occurrence that would have executed, so `skipPastOccurrences` may imply that if it's false that it might queue up multiple runs otherwise

not in love with the name `recomputeSchedule` either, so open to other suggestions

File: src/task/task.ts [hunk 4]
```diff
@@ -178,8 +195,9 @@ export interface Task<T> extends TaskBase<T> {
    * Enable the task for processing.
    *
    * @public
+   * @param options Job Enabling Options
```
Comment (princjef) [inline src/task/task.ts:198]:
```suggestion
   * @param options Task Enabling Options
```

File: src/types/public.ts (hunk 3 above)
Comment (shubhamms) [reply src/types/public.ts:67]: Yeah that sounds like a better name. skipPastOccurrences will probably confuse people on the negative case. Ill make that change along with other changes

File: src/task/task.ts [hunk 5]
```diff
@@ -101,12 +103,27 @@ export default class TaskImpl<T> implements Task<T> {
     );
   }
 
-  async enable(): Promise<void> {
+  async enable(options?: TaskEnablingOptions): Promise<void> {
+    let patchOp = async () => this._data.patch(() => ({ enabled: true }));
+
+    // Patch task with enabling options
+    if (options) {
+      if (!!options.skipPastOccurrences && this._data.nextRunTime) {
```
Comment (princjef) [inline src/task/task.ts:111]: can be collapsed into a single condition

```suggestion
    if (options && options.skipPastOccurrences && this._data.nextRunTime) {
    }
```

File: src/task/task.ts [hunk 6]
```diff
@@ -101,12 +103,27 @@ export default class TaskImpl<T> implements Task<T> {
...
+        const nextRunTime = computeNextRun(
+          this._data.interval,
+          this._data.nextRunTime.getTime(),
```
Comment (princjef) [inline src/task/task.ts:114]: wouldn't you want this to be based on the current time? otherwise, this could end up pushing your next run way out. if for instance your schedule was once per week and your next run is scheduled in 6 days, this would push it out to 13 days

File: src/types/public.ts [hunk 7]
```diff
@@ -55,3 +55,14 @@ export interface TimeoutsOptions {
    */
   retries?: number;
 }
+
+export interface TaskEnablingOptions {
```
Comment (princjef) [inline src/types/public.ts:59]: nit: i think alignment with type naming conventions for the other types would be `TaskEnableOptions` (don't actually remember though)

**Review by princjef — CHANGES_REQUESTED** [review]
Looks good overall, just a few comments. Thanks for including a test 💯

File: src/task/task.ts (hunk 6 above)
Comment (shubhamms) [reply src/task/task.ts:114]: sure makes sense.

Comment (shubhamms) [reply src/task/task.ts:114]: In that case I wouldnt have to change the code in computeNextRun and it will just base off of the current time. Ill revert those changes as well

**Comment by princjef** [comment]
should your added option be available on the `enable`/`enableAll` methods in the client as well?

**Review by princjef — APPROVED** [review]
LGTM
